### PR TITLE
pythonPackages.mypy: 0.812 -> 0.910

### PR DESCRIPTION
**Pull request by veehaitch**
###### Motivation for this change

Update `mypy` from version `0.812` to `0.910`.

Introduces new build-time dependencies on `types-toml` and `types-typed-ast` which are both added as Python modules. Unfortunately, building fails:

```
[...]
python3.8-mypy> mypy/modulefinder.py:436: error: Library stubs not installed for "toml" (or incompatible with Python 3.8)
python3.8-mypy> mypy/config_parser.py:10: error: Library stubs not installed for "toml" (or incompatible with Python 3.8)
python3.8-mypy> mypy/config_parser.py:10: note: Hint: "python3 -m pip install types-toml"
python3.8-mypy> mypy/config_parser.py:10: note: (or run "mypy --install-types" to install all missing stub packages)
python3.8-mypy> mypy/config_parser.py:10: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
python3.8-mypy> mypy/fastparse.py:109: error: Library stubs not installed for "typed_ast" (or incompatible with Python 3.8)
python3.8-mypy> mypy/fastparse.py:109: note: Hint: "python3 -m pip install types-typed-ast"
python3.8-mypy> mypy/fastparse2.py:57: error: Library stubs not installed for "typed_ast.ast27" (or incompatible with Python 3.8)
python3.8-mypy> mypy/fastparse2.py:68: error: Library stubs not installed for "typed_ast" (or incompatible with Python 3.8)
[...]
```

Building complains about missing stubs for `toml` and `typed-ast` although I added them to the `nativeBuildInputs`.

Maybe someone else has an idea why this is happening.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bjornfor** [comment]
Ref. https://github.com/nix-community/poetry2nix/issues/323.

**Comment by veehaitch** [comment]
Thanks to @bjornfor's reference, I can confirm that disabling `mypyc` results in a successful build. Given the (huge) runtime impact, it's not really an option though.

**Comment by veehaitch** [comment]
I disabled `mypyc` for versions newer than `v0.900` unless somebody knows how to make the derivation work with `mypyc` enabled.

**Comment by bjornfor** [comment]
Please reorder or squash the last two commits to prevent build breakage.

**Review by bjornfor — APPROVED** [review]

**Comment by wamserma** [comment]
Closes https://github.com/NixOS/nixpkgs/issues/139319
(so the issue gets closed when any one of the two mypy-update-PRs get merged)

File: pkgs/development/python-modules/mypy/default.nix [hunk 1]
```diff
@@ -1,18 +1,31 @@
-{ lib, stdenv, fetchPypi, buildPythonPackage, typed-ast, psutil, isPy3k
+{ lib
+, stdenv
+, fetchPypi
+, buildPythonPackage
+, pythonOlder
```
Comment (jnetod) [inline pkgs/development/python-modules/mypy/default.nix:5]:
```suggestion
, python
, pythonOlder
```

File: pkgs/development/python-modules/mypy/default.nix [hunk 2]
```diff
@@ -1,18 +1,31 @@
...
+, typed-ast
 , mypy-extensions
 , typing-extensions
+, toml
+, types-toml
+, types-typed-ast
 }:
 buildPythonPackage rec {
```
Comment (jnetod) [inline pkgs/development/python-modules/mypy/default.nix:13]:
```suggestion
let
  pythonEnv = python.withPackages (_: [ types-toml types-typed-ast ]);
in buildPythonPackage rec {
```

File: pkgs/development/python-modules/mypy/default.nix [hunk 3]
```diff
@@ -1,18 +1,31 @@
...
+    sha256 = "sha256-cECYMCRzyzGiGPF3Woc7N2swtMGCKUIenp3IkW/RYVA=";
   };
 
-  propagatedBuildInputs = [ typed-ast psutil mypy-extensions typing-extensions ];
+  nativeBuildInputs = [
+    types-toml
+    types-typed-ast
+  ];
+
+  propagatedBuildInputs = [ typed-ast mypy-extensions typing-extensions toml ];
```
Comment (jnetod) [inline pkgs/development/python-modules/mypy/default.nix:29]:
```suggestion

  postPatch = ''
    echo "python_executable = ${pythonEnv}/bin/python" >> mypy_bootstrap.ini
  '';
  
```

File: pkgs/development/python-modules/mypy/default.nix [hunk 4]
```diff
@@ -29,13 +42,16 @@ buildPythonPackage rec {
 
   # Compile mypy with mypyc, which makes mypy about 4 times faster. The compiled
   # version is also the default in the wheels on Pypi that include binaries.
-  # is64bit: unfortunately the build would exhaust all possible memory on i686-linux.
-  MYPY_USE_MYPYC = stdenv.buildPlatform.is64bit;
+  MYPY_USE_MYPYC =
+    # is64bit: unfortunately the build would exhaust all possible memory on i686-linux.
+    stdenv.buildPlatform.is64bit
+    # Derivation fails to build since v0.900 if mypyc is enabled.
+    && lib.strings.versionOlder version "0.900";
```
Comment (jnetod) [inline pkgs/development/python-modules/mypy/default.nix:49]:
```suggestion
    stdenv.buildPlatform.is64bit;
```

File: pkgs/development/python-modules/mypy/default.nix (hunk 3 above)
Comment (jnetod) [reply pkgs/development/python-modules/mypy/default.nix:29]: > If you have installed typed packages in another Python installation or environment, mypy won’t automatically find them. One option is to install another copy of those packages in the environment in which you use to run mypy. Alternatively, you can use the --python-executable flag to point to the target Python executable, and mypy will find packages installed for that Python executable.

https://github.com/python/mypy/blob/v0.910/docs/source/installed_packages.rst#using-installed-packages-with-mypy-pep-561

**Comment by mweinelt** [comment]
@veehaitch That went nowhere, are you interested in looking into this further?

**Comment by veehaitch** [comment]
I didn't quite follow the development in #140627. What's left to do? If it's just addressing the pending suggestions, I'm happy to reopen this PR.

**Comment by mweinelt** [comment]
Mypy is still stuck at 0.812. Not exactly sure what is missing. Does it find the library stubs?

**Pull request closed without merging**